qiskit/result/result.py

```diff
@@ -69,6 +69,7 @@ def __init__(
         if header is not None:
             self.header = header
         self._metadata.update(kwargs)
+        self.get_counts_type = None
```

Comment:
I think this might be better as a Boolean (with a suitable name - perhaps `force_list_from_get_counts`) rather than an arbitrary type - that's rather more self-documenting what it does.  As it stands, it's not clear what types actually have an effect.

It would also be a bit cleaner design from a behaviour-separability point of view if this was a parameter to `__init__`, rather than requiring to be modified after the fact.  That might require `QuantumInstance` to pass some more data through to its internal helper functions in this particular case, but it's generally more robust design if all an object's properties get set during its initialisation, rather than programmers being required to remember to mutate them in the correct ways later on.  People forget / don't know to do that / whatever, and things easily get lost in this style.